### ISSUE TITLE
Memory Allocations

### ISSUE THREAD

**Issue by Shadorain** [mattduck/kilo#1]
Hello mattduck! Thank you for having this repo up, it helped me in the early stages of my editor to find a way to go after completing the full kilo tutorial. One thing that I didn't realize until I was already quite far into was the memory ramifications of kilo as an entirety. 

Strings are not a valid data structure by any means for an extensive text editor, and Kilo I don't believe has that purpose in mind. But, as a tutorial it should point to methods of better achieving a valid way of structuring the data. 

I'm putting this valgrind log here just to show you how bad it is (simply by opening `./kilo kilo.c` from your repo) and running a few movement and `dd` commands. I do this as a warning because it wasn't until too far in that I realized and now am in the process of an overhaul of my entire back-end system of managing the data and went through weeks of research.

Hope this little issue helps you to progress your editor to be even better! Thanks!! If you would like to talk about ways I found and will go about my editor feel free to hit me up or write it in a reply!

`$> valgrind -s --log-file=./.valgrind.log --leak-check=full --show-leak-kinds=all --track-origins=yes ./kilo kilo.c`

```
==25676== Memcheck, a memory error detector
==25676== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==25676== Using Valgrind-3.16.1 and LibVEX; rerun with -h for copyright info
==25676== Command: ./kilo kilo.c
==25676== Parent PID: 23413
==25676== 
==25676== 
==25676== HEAP SUMMARY:
==25676==     in use at exit: 170,929 bytes in 4,176 blocks
==25676==   total heap usage: 599,703 allocs, 595,527 frees, 1,081,328,049 bytes allocated
==25676== 
==25676== 7 bytes in 1 blocks are still reachable in loss record 1 of 6
==25676==    at 0x483877F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==25676==    by 0x48ED6EA: strdup (strdup.c:42)
==25676==    by 0x10BF1C: editorOpen (in /home/shadow/Downloads/kilo/kilo)
==25676==    by 0x10E3F5: main (in /home/shadow/Downloads/kilo/kilo)
==25676== 
==25676== 128 bytes in 1 blocks are still reachable in loss record 2 of 6
==25676==    at 0x483877F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==25676==    by 0x10D251: editorPrompt (in /home/shadow/Downloads/kilo/kilo)
==25676==    by 0x10C535: editorColon (in /home/shadow/Downloads/kilo/kilo)
==25676==    by 0x10DE5D: editorProcessKeypressNormalMode (in /home/shadow/Downloads/kilo/kilo)
==25676==    by 0x10E421: main (in /home/shadow/Downloads/kilo/kilo)
==25676== 
==25676== 33,204 bytes in 1,391 blocks are still reachable in loss record 3 of 6
==25676==    at 0x48386AF: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==25676==    by 0x483ADE7: realloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==25676==    by 0x10A8D2: editorUpdateSyntax (in /home/shadow/Downloads/kilo/kilo)
==25676==    by 0x10B321: editorUpdateRow (in /home/shadow/Downloads/kilo/kilo)
==25676==    by 0x10B606: editorInsertRow (in /home/shadow/Downloads/kilo/kilo)
==25676==    by 0x10BFBA: editorOpen (in /home/shadow/Downloads/kilo/kilo)
==25676==    by 0x10E3F5: main (in /home/shadow/Downloads/kilo/kilo)
==25676== 
==25676== 34,595 bytes in 1,391 blocks are still reachable in loss record 4 of 6
==25676==    at 0x483877F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==25676==    by 0x10B4D0: editorInsertRow (in /home/shadow/Downloads/kilo/kilo)
==25676==    by 0x10BFBA: editorOpen (in /home/shadow/Downloads/kilo/kilo)
==25676==    by 0x10E3F5: main (in /home/shadow/Downloads/kilo/kilo)
==25676== 
==25676== 34,595 bytes in 1,391 blocks are still reachable in loss record 5 of 6
==25676==    at 0x483877F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==25676==    by 0x10B242: editorUpdateRow (in /home/shadow/Downloads/kilo/kilo)
==25676==    by 0x10B606: editorInsertRow (in /home/shadow/Downloads/kilo/kilo)
==25676==    by 0x10BFBA: editorOpen (in /home/shadow/Downloads/kilo/kilo)
==25676==    by 0x10E3F5: main (in /home/shadow/Downloads/kilo/kilo)
==25676== 
==25676== 68,400 bytes in 1 blocks are still reachable in loss record 6 of 6
==25676==    at 0x483AD7B: realloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==25676==    by 0x10B392: editorInsertRow (in /home/shadow/Downloads/kilo/kilo)
==25676==    by 0x10BFBA: editorOpen (in /home/shadow/Downloads/kilo/kilo)
==25676==    by 0x10E3F5: main (in /home/shadow/Downloads/kilo/kilo)
==25676== 
==25676== LEAK SUMMARY:
==25676==    definitely lost: 0 bytes in 0 blocks
==25676==    indirectly lost: 0 bytes in 0 blocks
==25676==      possibly lost: 0 bytes in 0 blocks
==25676==    still reachable: 170,929 bytes in 4,176 blocks
==25676==         suppressed: 0 bytes in 0 blocks
==25676== 
==25676== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```
`1,081,328,049 bytes allocated` <-- Thats an entire gigabyte+!! This is just unreasonable (even worse... try opening a massive file [warning your system might crash])

**Comment by mattduck** [comment]
@Shadorain hi! Thanks for writing this. Glad you found the repo (partially) useful. I hope the restructure wasn't too problematic.

I was aware that there are major memory issues in this code and that valgrind returns these errors. My guess is that they're not present in the original Kilo code, and that they're all issues that I introduced. Without looking into it I'd imagine that most of them relate to the very naive undo/redo implementation that I started which copies a lot of the editor state. But it's also possible there are other problems.

This repo was something fun that I did for a week last year and sadly I'm not likely to revisit it anytime soon. I hadn't considered that somebody might run into issues using this to build their own implementation, so thanks for raising that. I'll add a disclaimer to the README this afternoon.

Good luck with your editor! I'd definitely be interested in reading about the changes you made vs Kilo and where things break down in the original implementation.
